### ISSUE TITLE
[Proposal] Creating diffs to work in LLM distributed settings

### ISSUE THREAD

**Issue by Narsil** [huggingface/safetensors#242]
# Inference

Currently for inference we make extensive use of the lazy loading in order to be able to deploy efficiently TP sharded models with the same set of weights:

Like here: 

https://github.com/huggingface/text-generation-inference/blob/main/server/text_generation_server/models/flash_llama.py#L216-L311

# Training situation

However the situation is not yet really good for training LLMs, where we have even more sharding, more state to save (like optimizer states) across various nodes.

With setups like DS Zero3 there is sharding TP and potentially PP at the same, meaning a single process can have only part of the weights at a given time.

Currently afaik it involves many tricks in order to save process local weights, and reloading on a different topology (reloading on the same topology is always easy, just reloaded whatever the previous process in the same location had saved) is quite tricky and very code involved.

However reloading on a different topology is really interesting in order to try and maximize the total FLOPs of a training. Like more sharding to allow for bigger batch size.

## Constraints

- **Avoid network code inside the lib**: I really want to avoid coding networking things into `safetensors`. It makes code more complex, SSL handling is always a hassle to distribute "correctly" and "hassle-free": https://github.com/huggingface/tokenizers/issues/1113

- **Be network friendly**: In a distributed setup, using GCP, S3 and other bucket like storage is a common strategy which alleviates the need to care too much about where files are. The thing is that while we can do Partial reads on those (using `Byte-Range` header when issueing GET requests). We cannot do Partial Writes in the same way. https://stackoverflow.com/questions/34761076/can-we-update-contents-of-a-specific-file-in-amazon-s3 . For the very large files handled by the large models, it's very painful to need to reupload entire files whenever an update is made.

- **Be able to store arbitrary slices**: As said before any process stores arbitrary slices (TP) of the full model so we need to be able to store these. In a single node setting we could just write the appropriate places of the data buffer.

## Proposal: Create DIFF format for safetensors

This would be something very close to `safetensors` format, except it would describe how to *modify* and existing model file with given diff data. (A bit similar to what git does with diff files)

It's  basically like saving the current process state_dict of a given process (like what is done currently) except it would keep enough information to be reusable by different process at a later time. Something like this:

```python
# Training process on node 1 rank 0
with create_diff("model_node1_rank0.diff", framework="pt") as f:
    f.get_slice("layer.0.attn.qkv.weight")[:, :1600] = self.layer[0].attn.qkv.weight
    f.get_slice("layer.0.attn.qkv.bias")[:] = self.layer[0].attn.qkv.bias
    
    
# Later on a fusing process could look like

filename = "model-0001-of-0007.safetensors"
create_empty_file({"layer.0.attn.qkv.weight": {"shape": [1600, 4800], "dtype": "F32"}, "layer.0.attn.qkv.bias": {"shape": [1600], "dtype": "F32"}, filename}

with edit(filename) as f:
    with patch_open("model_node1_rank0.diff") as g:
        for patch in g:
            if patch["name"] in f.keys():
                f.apply(patch)
```

We could envision different APIs here, the key thing is:
- A single process handles the merging. Writing concurrently to files is error prone. The `edit` function could take care of checking that no overlapping writes happens for instance
- The final layout of the weights (split across n files or single file) is agnostic to the diff, diffs can be applied independently to parts of weights files because the patch information can keep information about where the patch is supposed to be applied.
- All diff files can be uploaded at any given point to S3 in whatever order (like `checkpoint-500/xxxx.diff`) and each process uploads a different file, there's no issue.
- This keeps `safetensors` network agnostic

Note:
- The merging of diffs on S3 would still mean downloading N diffs files, creating M new (possibly sharded) safetensors file and uploading them, which can take some time. However the resulting files can indeed be loaded with any topology, and because of the S3 limitation of no partial updates, there's no real way to do a better job.
- Big con is adding a new file format. But git does use `diff` and regular files in a similar fashion giving me confidence it should be OK.

## Things open to discussion.

Error handling:

- Can single process create ANY diff file without knowing anything about the fully fledged model ? Or should we enforce knowing the entire model structure, to make sure the proposed diffs are VALID against it. (Currently I think making validation optional would be OK)
- What happens on applying 2 diffs that override the same part of the file. This is likely an error that users would like to be warned about. Maybe there needs to be several merging strategies for things like num_batched_processed in the case of batchnorm/layernorm where values need to be averaged over.
- What about conflicts ? Since the diff files only intends to capture modifying the databuffer, not the overall structure the only kind of merge issue is the one mentioned above so we shouldn't have the need to do things like *rebase*
